lib/checkclass.cpp

```diff
@@ -1551,7 +1565,42 @@ bool CheckClass::hasAllocation(const Function *func, const Scope* scope) const
     return false;
 }
 
-bool CheckClass::hasAssignSelf(const Function *func, const Token *rhs)
+bool CheckClass::isInverseAssignmentTest(const Token *tok)
+{
+	bool res = true;
+	for (const Token *itr = tok; itr; itr=itr->astParent()) {
+		if (itr->str() == "!=" && (itr->astOperand1()->str() == "true" || itr->astOperand2()->str() == "true")) {
+			res = !res;
+		}
+		else if (itr->str() == "!=" && (itr->astOperand1()->str() != "false" && itr->astOperand2()->str() != "false")) {
+			res = !res;
+		}
+		else if (itr->str() == "!") {
+			res = !res;
+		}
+		else if (itr->str() == "==" && (itr->astOperand1()->str() == "false" || itr->astOperand2()->str() == "false")) {
+			res = !res;
+		}
+	}
+	return res;
+}
+
+const Token * CheckClass::getScopeStartToken(const Token *tok)
```

Comment:
hmm .. this is a strange name. With this name I would have thought it worked like `tok->scope()->bodyStart`.

I guess I'd prefer that you use `tok->astTop()` instead of a for loop. I assume that we only are interested in the top.

Then instead of such condition: `itr->link()->next() && itr->link()->next()->str() == "{"` I'd prefer `Token::simpleMatch(itr->link(), ") {")`. That is more explicit and also avoid extra calls for link() and next().

Reply:
I think the name is strange but do not have a suggestion right now.. feel free to suggest something..

Reply:
Done